streamingest: add custom ctx to support cutting over

adityamaru opened the pull request:
This change adds logic to leverage context cancelation as a means of
trigerring the RevertRange logic that is run when the job is requested
to cutover.

This change also adds a poller to check if the job has been marked for
completion.

Release note: None

pbardea committed:
streamingest: add ctx cancelation to support cutting over

This change adds logic to leverage context cancelation as a means of
trigerring the RevertRange logic that is run when the job is requested
to cutover.

This change also adds a poller to check if the job has been marked for
completion.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60536)
<!-- Reviewable:end -->

adityamaru commented:
@pbardea the second commit has the relevant changes. The first commit is from https://github.com/cockroachdb/cockroach/pull/59588.

adityamaru commented:
Closing in favour of #60763.

adityamaru committed:
streamingest: add OnFailOrCancel ClearRange logic to stream ingestion

When the job fails, we must ClearRange the target keyspace being
ingested into.

Release note: None

adityamaru commented:
`TestStreamIngestionJobWithRandomClient` is currently skipped cause it takes ~30seconds but I unskipped it while testing the custom ctx cancelation. The majority of the time is spent waiting for the job to transition to the `Canceled` state so wondering if y'all have any ideas to restructure this?

adityamaru commented:
Closing in favour of https://github.com/cockroachdb/cockroach/pull/61001, Reviewable doesn't work on this PR.
